File: seg/Preprocessing/script_bench.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import cv2
import numpy as np
import os
from PIL import Image
from time import time

from shad_atten import shad_atten as shad_atten

# EDIT HERE (1 of 2):
verbose = 1

# EDIT HERE (2 of 2):
#dataset_name = 'DermNet'
#dataset_name = 'DermIS'
#dataset_name = 'DermQuest'
#dataset_name = 'MED-NODE'
#dataset_name = 'PAD-UFES'
#dataset_name = 'MClass-ND'
dataset_name = 'MClass-D'

# Setting the source and destination dirs:
src_dir_root = os.path.join('/home/eliezer/Área de Trabalho/thesis', dataset_name)
src_dir_melanoma = os.path.join(src_dir_root, 'melanoma')
src_dir_notmelanoma = os.path.join(src_dir_root, 'notmelanoma')

# Creating the destination dirs, if they do not exist:
curr_dir = os.getcwd()
if verbose == 0:  
    dst_dir_root = os.path.join('/home/eliezer/Área de Trabalho/thesis', dataset_name + '_preprocessed')
else:
    dst_dir_root = os.path.join('/home/eliezer/Área de Trabalho/thesis/seg/Preprocessing', dataset_name + '_preprocessed')
dst_dir_melanoma = os.path.join(dst_dir_root, 'melanoma')
dst_dir_notmelanoma = os.path.join(dst_dir_root, 'notmelanoma')
if not os.path.exists(dst_dir_root):
    os.mkdir(dst_dir_root)
    os.mkdir(dst_dir_melanoma)
    os.mkdir(dst_dir_notmelanoma)

elapsed_times = []

# Preprocessing all the images of the selected dataset.
for src_dir, dst_dir in zip([src_dir_melanoma, src_dir_notmelanoma], [dst_dir_melanoma, dst_dir_notmelanoma]):
    
    fnames_tmp = os.listdir(src_dir)
    fnames = [fname for fname in fnames_tmp if fname.endswith('.jpg')]    
    nimgs = len(fnames)
    
    for i in range(nimgs):
    
        print('Working on {:s} image {:3d} of {:3d}...'.format(dst_dir.split('/')[-1], i + 1, nimgs))
        
        name = fnames[i].split('.')[0]

        img_tmp = cv2.imread(os.path.join(src_dir, name + '.jpg'))
        img = cv2.cvtColor(img_tmp, cv2.COLOR_BGR2RGB)
        if dataset_name in ['DermNet','DermIS','DermQuest']:
            gt_tmp = cv2.imread(os.path.join(src_dir, name + '.png'))
            gt = cv2.cvtColor(gt_tmp, cv2.COLOR_BGR2GRAY)
            
        ti = time()
                
        nr = img.shape[0]
        nc = img.shape[1]
        
        kappa = np.int(0.2*min(nr,nc))
        mask = np.zeros((nr, nc), dtype=int)
        mask[0:kappa, 0:kappa] = 1
        mask[0:kappa, nc-kappa:nc] = 1
        mask[nr-kappa:nr, 0:kappa] = 1
        mask[nr-kappa:nr, nc-kappa:nc] = 1
                
        saturate_opt = 1
        normalize_opt = 1    
        
        if verbose == 0:
        
            img_corr, _, _, _ = shad_atten(img, mask, saturate_opt, normalize_opt)    
            
            img_pil = Image.fromarray(img_corr, 'RGB')
            
            dt = time() - ti
            elapsed_times.append(dt)
            
            img_pil.save(os.path.join(dst_dir, name + '.jpg'))
            
            if dataset_name in ['DermNet','DermIS','DermQuest']:
                gt_pil = Image.fromarray(gt, 'L')
                gt_pil.save(os.path.join(dst_dir, name + '.png'))            
            
        else:
        
            img_pil = Image.fromarray(img, 'RGB')
            img_pil.save(os.path.join(dst_dir, name + '_orig.jpg'))        
        
            img_corr, Vin, Vout, Z = shad_atten(img, mask, saturate_opt, normalize_opt)
        
            imgcorr_pil = Image.fromarray(img_corr, 'RGB')
            imgcorr_pil.save(os.path.join(dst_dir, name + '_corr.jpg'))
    
            Vin_pil = Image.fromarray(Vin, 'L')
            Vin_pil.save(os.path.join(dst_dir, name + '_Vorig.jpg'))
            
            Vout_pil = Image.fromarray(Vout, 'L')
            Vout_pil.save(os.path.join(dst_dir, name + '_Vcorr.jpg'))
            
            Z_pil = Image.fromarray(Z, 'L')
            Z_pil.save(os.path.join(dst_dir, name + '_Z.jpg'))
     
np.savetxt(os.path.join(dst_dir_root, 'elapsed_time.csv'), elapsed_times, delimiter=',')            

File: seg/Segmentation/script_bench.m
clear all;
close all;
clc;

addpath('src');
addpath('src/Ncut_9');

% EDIT HERE (1 of 4)
verbose = 0;

% EDIT HERE (2 of 4)
dataset_name = 'MClass-D';

% EDIT HERE (3 of 4)
%for K = 3:9
for K = 3:3

    % EDIT HERE (4 of 4)
    %for alternative_method = 1:4
    for alternative_method = 1:1

        fprintf('* With K = %d and alternative_method = %d\n', K, alternative_method);

        curr_dir = pwd;

        %% Setting the source and destination dirs:

        if alternative_method == 1
            src_dir_root = ['/home/eliezer/Área de Trabalho/thesis/' dataset_name];
        else
            src_dir_root = ['/home/eliezer/Área de Trabalho/thesis/' dataset_name '_preprocessed'];
        end
        src_dir_melanoma = [src_dir_root '/melanoma'];
        src_dir_notmelanoma = [src_dir_root '/notmelanoma'];

        if alternative_method == 0
            dst_dir_root = [curr_dir '/' dataset_name '_K' num2str(K)];
        else
            dst_dir_root = [curr_dir '/' dataset_name '_K' num2str(K) '_A' num2str(alternative_method)];
        end
        dst_dir_melanoma = [dst_dir_root '/melanoma'];
        dst_dir_notmelanoma = [dst_dir_root '/notmelanoma'];

        if ~exist(dst_dir_root, 'dir')
            mkdir(dst_dir_root);
            mkdir(dst_dir_melanoma);
            mkdir(dst_dir_notmelanoma);
        end

        if verbose == 1

            if alternative_method == 0
                dst_dir_details_root = [curr_dir '/' dataset_name '_K' num2str(K) '_details'];
            else
                dst_dir_details_root = [curr_dir '/' dataset_name '_K' num2str(K) '_A' num2str(alternative_method) '_details'];
            end
            dst_dir_details_melanoma = [dst_dir_details_root '/melanoma'];
            dst_dir_details_notmelanoma = [dst_dir_details_root '/notmelanoma'];

            if ~exist(dst_dir_details_root, 'dir')
                mkdir(dst_dir_details_root);
                mkdir(dst_dir_details_melanoma);
                mkdir(dst_dir_details_notmelanoma);
            end

        end

        %% Segmenting all the images of the selected dataset.

        dirs_cell = cell(3,2);
        dirs_cell{1,1} = src_dir_melanoma;
        dirs_cell{1,2} = src_dir_notmelanoma;
        dirs_cell{2,1} = dst_dir_melanoma;
        dirs_cell{2,2} = dst_dir_notmelanoma;

        if verbose == 1
            dirs_cell{3,1} = dst_dir_details_melanoma;
            dirs_cell{3,2} = dst_dir_details_notmelanoma;
        end

        elapsed_times = [];
        for dirs = dirs_cell

            src_dir = dirs{1};
            dst_dir = dirs{2};

            if verbose == 1
                dst_dir_details = dirs{3};
            end

            cd(src_dir);
            dir_struct = dir('*.jpg');
            cd(curr_dir);
            nimgs = numel(dir_struct); 

            for i = 1:nimgs

                src_dir_str_parts = split(src_dir, '/');
                fprintf('Segmenting %s image %d of %d...\n', src_dir_str_parts{end}, i, nimgs);

                name = dir_struct(i).name(1:end-4);
                img = imread([src_dir '/' name '.jpg']);

                tic;

                nr = size(img, 1);
                nc = size(img, 2);

                w = 2*round((0.01*max(nr, nc)+1)/2) - 1;

                if alternative_method == 2

                    seg = dict_segmentation_A2(img, w, K); 

                elseif alternative_method == 3

                    addpath('src/ompbox10');
                    addpath('src/ksvdbox13');                    
                   
                    seg = dict_segmentation_A3(img, w, K); 

                elseif alternative_method == 4
                    
                    addpath(genpath('src/vlfeat-0.9.20'));
                    run('vl_setup');
                    
                    seg = dict_segmentation_A4(img, w, K); 
                    
                else

                    if verbose == 0

                        seg = dict_segmentation(img, w, K);

                    else

                        [seg, assessed_rectangles, optimal_rectangle, D0] = dict_segmentation(img, w, K);

                        for r = 1:length(assessed_rectangles)    
                            if sum(sum(assessed_rectangles{r})) == sum(sum(optimal_rectangle))
                                imwrite(imoverlay(img, assessed_rectangles{r}, [0 0 1]), [dst_dir_details '/' name '_rect' num2str(r) '.png']);
                            else
                                imwrite(imoverlay(img, assessed_rectangles{r}, [0 0 0]), [dst_dir_details '/' name '_rect' num2str(r) '.png']);
                            end
                        end

                        print_dict(D0, [dst_dir_details '/' name '_D0.png']);

                    end

                end

                elapsed_time = toc;
                elapsed_times = [elapsed_times elapsed_time];

                imwrite(seg, [dst_dir '/' name '.png']);

            end

        end

        csvwrite([dst_dir_root '/elapsed_time.csv'], elapsed_times);

    end % End for 'alternative_method'.
    
end % End for 'K'.
